crates/wasm-smith/tests/exports.rs

```diff
@@ -0,0 +1,246 @@
+        )"#,
+        expected_exports: vec![
+            make_global("1i32", ValType::I32, true),
+            make_global("2i32", ValType::I32, true),
+            make_global("1i64", ValType::I64, true),
+            make_global("2i64", ValType::I64, true),
+            make_global("3i32", ValType::I32, true),
+            make_global("3i64", ValType::I64, true),
+            make_global("4i32", ValType::I32, false),
+            make_global("4i64", ValType::I64, false),
+        ],
+    };
+    smoke_test_exports(test, 22)
+}
+
+fn smoke_test_exports(exports_test_case: TestCase, seed: u64) {
+    let mut rng = SmallRng::seed_from_u64(seed);
+    let mut buf = vec![0; 512];
+
+    for _ in 0..1024 {
+        rng.fill_bytes(&mut buf);
+        let mut u = Unstructured::new(&buf);
+
+        let mut config = Config::arbitrary(&mut u).expect("arbitrary config");
+        config.exports = Some(wat::parse_str(&exports_test_case.export_module).unwrap());
+
+        let features = parser_features_from_config(&config);
+        let module = Module::new(config, &mut u).unwrap();
+
+        let wasm_bytes = module.to_bytes();
+        let mut validator = Validator::new_with_features(features);
+        validate(&mut validator, &wasm_bytes);
+
+        let mut types = vec![];
+        let mut func_imports = vec![];
+        let mut global_imports = vec![];
+        let mut funcs = vec![];
+        let mut globals = vec![];
+        let mut exports = vec![];
+
+        for payload in Parser::new(0).parse_all(&wasm_bytes) {
+            let payload = payload.unwrap();
+            if let wasmparser::Payload::TypeSection(rdr) = payload {
+                // Gather the signature types to later check function types
+                // against.
+                for ty in rdr.into_iter_err_on_gc_types() {
+                    types.push(ty.unwrap());
+                }
+            } else if let wasmparser::Payload::FunctionSection(rdr) = payload {
+                funcs = rdr.into_iter().collect::<Result<_, _>>().unwrap();
+            } else if let wasmparser::Payload::ImportSection(rdr) = payload {
+                for import in rdr {
+                    let import = import.expect("Failed to read import");
+                    match import.ty {
+                        TypeRef::Func(i) => func_imports.push(i),
+                        TypeRef::Global(g) => global_imports.push(g),
+                        TypeRef::Table(_) | TypeRef::Memory(_) | TypeRef::Tag(_) => {}
+                    }
+                }
+            } else if let wasmparser::Payload::GlobalSection(rdr) = payload {
+                for global in rdr {
+                    let global = global.expect("Failed to read global");
+                    globals.push(global.ty);
+                }
+            } else if let wasmparser::Payload::ExportSection(rdr) = payload {
+                exports = rdr.into_iter().collect::<Result<_, _>>().unwrap();
+            }
+        }
```

Comment:
One thing I might recommend is to use the [result of validation](https://docs.rs/wasmparser/latest/wasmparser/types/struct.Types.html) which should avoid the need to manually walk over these sections and record the results.

Reply:
You might also be able to skip the `expected_exports` field and instead run the validator over the `export_module` field and then assert everything there is also present in the wasm-smith output

Reply:
Thanks, I switched to using `Validator` to get `Types` in both the test and the implementation.